### ISSUE TITLE
Issues sending multiple messages with HighLevelProducer

### ISSUE THREAD

**Issue by mdnorman** [SOHU-Co/kafka-node#515]
We are having trouble sending multiple messages to the same topic & partition with the HighLevelProducer. It's only sending the first message in the partition.

Here's our setup:
- Using **HighLevelProducer** with the **keyed** partitioner type
- Sending multiple payloads in an array to the **send** function
- Each payload has a topic, a key, a single **KeyedMessage**, and a compression attribute

Because we are using the keyed partitioner type, and the keys for each message are unique, we must create a single payload per key instead. The HighLevelProducer **send** function determined the partition based on the partitioner, so there could be multiple payloads with the same topic and partition. The HighLevelProducer uses **Client** to **sendProduceRequests** with an array of payloads, which seems reasonable.

The Client then splits the payloads by broker based on leader per partition and uses the **protocol.encodeProduceRequest** to encode the list of payloads per broker.

The protocol _encodeProduceRequest immediately calls **groupByTopic** to split the payloads into a map of topics -> partitions -> payloads. This is where the problem arises. The issue is that **groupByTopic** assumes at this time that there is only one payload per topic partition, which isn't the case in this instance.

Therefore, the issue is either in **protocol.groupByTopic** because it should create an array of payloads per partition OR the issue is in **HighLevelProducer.buildPayloads** because it should be grouping the individual payloads by partition.

Unfortunately, we are unable to group them ourselves because we don't know the partition beforehand as that logic is in the HighLevelProducer.

Hopefully this is clear enough to be able to get the issue fixed. Please let me know if there's anything that needs further explanation.

**Comment by hyperlink** [comment]
Could you create small test case to reproduce this?

Sounds like the issue could be fixed by updating the `buildPayloads` method to normalize the payload by topic partition and attributes.

**Comment by mdnorman** [comment]
```
'use strict';

const kafkaNode = require('kafka-node');

const partitionerTypes = kafkaNode.HighLevelProducer.PARTITIONER_TYPES;
const theTopic = 'test.topic';

const compressionTypes = {
  none: 0,
  gzip: 1,
  snappy: 2,
};

const kafkaClient = new kafkaNode.Client('localhost:2181', 'test-consumer-group');

const theMessages = [
  { id: 'A', data: 1 },
  { id: 'B', data: 2 },
  { id: 'C', data: 3 },
  { id: 'D', data: 4 },
  { id: 'E', data: 5 },
];

const updateMetadataForTopic = (topic, cb) => {
  kafkaClient.loadMetadataForTopics([topic], (err, metadata) => {
    if (err) {
      cb(err);
      return;
    }

    kafkaClient.updateMetadatas(metadata);
    cb();
  });
};

const kafkaProducer = new kafkaNode.HighLevelProducer(kafkaClient, { partitionerType: partitionerTypes.keyed });

const sendMessages = (topic, messages, cb) => {
  const payloads = messages.map(message => ({
    topic,
    key: message.id,
    messages: [new kafkaNode.KeyedMessage(message.id, JSON.stringify(message))],
    attributes: compressionTypes.snappy,
  }));

  kafkaProducer.send(payloads, cb);
};

kafkaClient.on('ready', () => {
  console.log('Kafka client ready.');
});

kafkaClient.on('error', err => {
  console.error(`Kafka client error: ${err.stack || err.message}`, err);
  kafkaClient.close();
});

kafkaProducer.on('ready', () => {
  console.log(`Kafka producer ready.`);

  // required when using keyed partitioner for some reason
  updateMetadataForTopic(theTopic, err => {
    if (err) {
      console.error(`Error updating metadata for topic: ${err.stack || err.message}`, err);
      kafkaClient.close();
      return;
    }

    sendMessages(theTopic, theMessages, err2 => {
      if (err2) {
        console.error(`Error sending messages: ${err2.stack || err2.message}`, err2);
        kafkaClient.close();
        return;
      }

      console.log(`Sent ${theMessages.length} messages`);
      kafkaClient.close();
    });
  });
});
```

**Comment by mdnorman** [comment]
Added the test above.

Also notice that the metadata for a topic needs to be updated on the client before you can use the keyed partitioner. this is something we get around by just doing the update ourselves, but it was kind of tough to find the issue there.

**Comment by hyperlink** [comment]
I think the partitioners need the total number of partitions in a topic in order to calculate the partition to assign and that's retrieved from a call to `client.refreshMetadata()`.

So I took a stab at fixing this. Maybe you point your `package.json` to `SOHU-Co/kafka-node#payload-normalize` and see if the issue persists or if you find any new issues? Basically my change just tries to normalize the `ProduceRequest` by topic partition.

**Comment by mdnorman** [comment]
Thank you so much for the really quick turnaround. I've tested this in the code path that was breaking before and it appears that this is exhibiting the correct behavior. After looking at the diff, I think this solution makes a lot of sense.

**Comment by hyperlink** [comment]
Thank you for testing. I will work on getting this published soon.

**Comment by hyperlink** [comment]
published in `1.0.7`
